src/world/world_trait.rs

```diff
@@ -32,24 +30,8 @@ use std::fmt::Debug;
 /// game-specific state you need. The `Stepper::step` implementation for such a world would
 /// typically invoke `DefaultMechanicalWorld::step`, and any other game-specific update logic.
 pub trait World: Stepper + Default + Send + Sync + 'static {
-    /// The command that can be used by the game and the player to interact with the physics
-    /// simulation. Typically, this is an enum of some kind, but it is up to you.
-    type CommandType: Command;
-
-    /// The subset of state information about the world that can be used to fully recreate the
-    /// world. Needs to be serializable so that it can be sent from the server to the client. There
-    /// is nothing stopping you from making the [`SnapshotType`](World::SnapshotType) the same type
-    /// as your [`World`] if you are ok with serializing the entire physics engine state. If you
-    /// think sending your entire [`World`] is a bit too heavy-weighted, you can hand-craft and
-    /// optimise your own `SnapshotType` structure to be as light-weight as possible.
-    type SnapshotType: Debug + Clone + Serialize + DeserializeOwned + Send + Sync + 'static;
-
-    /// The subset of state information about the world that is to be displayed/rendered. This is
-    /// used by the client to create the perfect blend of state information for the current
-    /// rendering frame. You could make this [`DisplayState`](World::DisplayStateType) the same
-    /// structure as your [`World`] if you don't mind CrystalOrb making lots of copies of your
-    /// entire [`World`] structure and performing interpolation on all your state variables.
-    type DisplayStateType: DisplayState;
+    /// Configuration parameters that tweak how CrystalOrb works.
+    type ConfigType: Config;
```

Comment:
The sore point here is Command which is coupled both with World (via validate) and NetworkResource (via client_id).
It does not belong wholly to World nor to Network, thus the idea with putting it into separate thing - Config.

I moved all configuration to Config for ease of use. But I do see value in keeping World-only things in World for documentation purpose.

I agree with **B** as it shows that Client depends on both World and Network (via Config).
I will reshuffle stuff to remove world/network dependencies.

Reply:
> The sore point here is Command which is coupled both with World (via validate) and NetworkResource (via client_id).

Good point. Do you think the rest of the `World` should also depend on this `client_id` the same way as `Command`, or nah not really? I just remembered that, in another project I've been working on, I've been using `client_id` to [identify which rigid body belongs to which player](https://github.com/ErnWong/dango-tribute/blob/c0fd1ee1ff332fc35705ad3fa426f8b97b14518a/shared/src/physics_multiplayer.rs#L172-L186), for example, which helped me apply the right command to the right rigid body.

I think, if all those `World`, `Command`, `Snapshot` and `DisplayState` were to depend on `client_id`, it might be nice to lump them together underneath `World`. Since they only depend on *some way* to identify clients/players, and otherwise they don't actually need to know about the `NetworkResource`, I was thinking that maybe `World` could could be parameterised by this `ClientId`, and only make the connection between `ClientId` and `ConnectionHandleType` in the `Config` level. Thoughts on this?

> I moved all configuration to Config for ease of use. But I do see value in keeping World-only things in World for documentation purpose.

True, I can see how both ways have their merits :sweat_smile:

Reply:
> I think, if all those `World`, `Command`, `Snapshot` and `DisplayState` were to depend on `client_id`, it might be nice to lump them together underneath `World`. Since they only depend on _some way_ to identify clients/players, and otherwise they don't actually need to know about the `NetworkResource`, I was thinking that maybe `World` could could be parameterised by this `ClientId`, and only make the connection between `ClientId` and `ConnectionHandleType` in the `Config` level. Thoughts on this?

Nice catch. The point here is that both Network and World has a relation to external entity - Client. Both need a common way of identifying clients, which we should share via Config.
i will try sharing only that and see what happens. 😸

Reply:
> I will reshuffle stuff to remove world/network dependencies.

Unfortunately after a long struggle I don't see a way of decoupling these.
`WorldType::ConfigType::ConnectionHandleType` is not the same as `WorldType::ConfigType::ConnectionHandleType` and compilation fails.

I don't see a way of making these the same type without putting everything in the same trait.

Reply:
Darn, that's a shame, but thank you for taking your time to attempt it anyway.

(If you haven't deleted/undo'ed your refactor attempt but haven't committed yet, I think it'll be awesome to have it committed (even if it'll be reverted in the next commit) to record your efforts for future curious readers)

Reply:
I haven't committed it, as it was going nowhere 😉